### ISSUE TITLE
Is it possible to `unblock` a blocked memory desc?

### ISSUE THREAD

**Issue by pinzhenx** [intel/mkl-dnn#575]
For instance, given a `dnnl_OIhw2o8i8o2i` memory desc, is it possible to compute the strides, dims, padded_dims of an equivalent `dnnl_oihw` memory desc without comparing with all non-blocking desc in a pattern matching fashion?
Or because there are irreversible functions inside `fill_blocked`, there is no way to do that?

**Comment by emfomenk** [comment]
Hi @pinzhenx,

Not sure I completely understand the question.

If you are asking whether you can recover `OIhw2o8i8o2i` format tag based on the memory descriptor (that before was created with `dnnl_memory_desc_init_by_tag(ndims, dims, data_type, OIhw2o8i8o2i)`) the answer is it is possible, but quite difficult. You would need to investigate the [blocking_desc_t](https://github.com/intel/mkl-dnn/blob/4d6dcf9884c38398fd7de68896c191add354187d/include/dnnl_types.h#L788-L801) along with [memory_desc_t](https://github.com/intel/mkl-dnn/blob/4d6dcf9884c38398fd7de68896c191add354187d/include/dnnl_types.h#L882-L933) structures to understand the physical layout of tensor and match it with `OIhw2o8i8o2i`.

Related issues: #470, #556.

However if you initialized a memory descriptor with `OIhw2o8i8o2i` format tag and now want to find out the `dims` and `padded_dims` you can simply get them from the memory descriptor itself:
``` c++
    memory_desc_t md;
    memory_desc_init_by_tag(dims={o, i, h, w}, dt, OIhw2o8i8o2i);

    int64_t O = duv_up(o, 16), I = div_up(i, 16);
    assert(md.dims == {o, i, h, w});
    assert(md.padded_dims == O*16, I*16, h, w});
```
Getting strides is a little be more complicated, since there is no single stride for `o` dimension since it is blocked. DNNL library keeps the outermost strides explicitly, while the blocked dimensions are assumed to be dense, hence no need in keep the strides. For the `md` above:
``` c++
    dims_t expected_strides = {
        I * h * w * 2 * 8 * 8 * 2,
            h * w * 2 * 8 * 8 * 2,
                w * 2 * 8 * 8 * 2,
                    2 * 8 * 8 * 2};
    assert(md.format_desc.blocking.strides == expected_strides);
```

---
Useful links:
- [**[RFC] Memory descriptor**](https://github.com/intel/mkl-dnn/blob/rfc-api-changes-v1.0/doc/rfc/api-v1.0/rfc_memory_desc.md#rfc-memory-descriptor)
- [**Understanding Memory Formats**](https://intel.github.io/mkl-dnn/understanding_memory_formats.html)

**Comment by pinzhenx** [comment]
@emfomenk Thanks for your detailed response.

What I'm looking for is basically a general function that transforms any non-plain mem desc (at least for those come with pre-defined tags) to a plain one.

For mem desc created from a non-permuted tag like `aBCde16b16c`, it's easy to calculate the strides for its `abcde` counterpart. However, it seems difficult to recover the strides for a mem desc generated by `aCBde16b16c`, as we have no idea what's the previous dim.

**Comment by emfomenk** [comment]
Hmm, I am still confused...

No matter whether the tag is `aBC...` or `aCB...` the strides and dimensions are always go in order of `abc...`. I.e. the strides/dims/padded_dims are always indexed by logical dimensions which come in natural order `abcdef....`, while tags like `aBCde16b16c` or `aCBde16b16c` are helpers that define the physical memory layout.

``` c++
memory_desc_t md_abc, md_acb;
memory_desc_init_by_tag(md_abc, {a, b, c}, dt, dnnl_abc); // abc array in abc-format 
memory_desc_init_by_tag(md_acb, {a, b, c}, dt, dnnl_acb); // same abc array, but in acb-format

assert(md_abc.dims[:] == md_acb.dims[:]);
assert(md_abd.padded_dims[:] = md_acb.padded_dims[:]);
assert(md_abc.format_desc.blocking.strides[:] == {
    b*c, // stride for the logical dim a
    c,   // stride for the logical dim b
    1    // stride for the logical dim c
});
assert(md_acb.format_desc.blocking.strides[:] == {
    b*c, // stride for the logical dim a
    1,   // stride for the logical dim b, even though the format is acb !!!
    b    // stride for the logical dim c, even though the format is acb !!!
});
```

If this doesn't answer your question, could you please share the API of the function you are looking for and few examples that demonstrate what it should do in different situations? How it can be used would be very helpful as well.

**Comment by pinzhenx** [comment]
I've already come up with an implementation of it based on the explanation. Sorry for the confusion and thanks again for your answer.

**Comment by pinzhenx** [comment]
The API is like 
```
desc unblock(desc blocked_desc)
```

For example, if the input mem desc is defined by `dnnl_aCBde16b16c` with A=1, B=8, C=4, D=2, E=16
```
                A       B       C       D       E
dims            1       8       4       2       16
padded_dims     1       16      16      2       16
strides         8192    8192    8192    4096    256

inner_idxs      1       2
inner_blks      16      16
```


The output should be like
```
                A       B       C       D       E
dims            1       8       4       2       16
padded_dims     1       8       4       2       16
strides         1024    32      256     16      1

inner_idxs
inner_blks
```


If I know this mem desc comes from `acbdexxxxx`, I could easily recover its strides to a plain one.
```
blocked_desc.strides = {
    C*B*d*e*16*16,
        d*e*16*16,
      B*d*e*16*16,
          e*16*16,
          1*16*16
}

=>

new_desc.strides = {
    C*B*d*e*16*16/C/B*c*b,
        d*e*16*16,
      B*d*e*16*16/B*b,
          e*16*16,
          1*16*16
}
```

However, if we do not record its format tag or permutation info in the mem desc, I cannot do the reversion shown above. For example, if I assume the mem desc is non-permuted, I would recover it to something like
```
new_desc.strides = {
    C*B*d*e*16*16/C/B*c*b,
        d*e*16*16/C*c,
      B*d*e*16*16,
          e*16*16,
          1*16*16
}
```
which is not supposed to be. So does that mean I have to record the format tag or permutation info by myself?

**Comment by emfomenk** [comment]
Hi @pinzhenx,

Yeah, for the function you described it seems you indeed need to keep the original tag.

However, to me that seems quite similar to the following situation: say you reordered the data from `acb` format to `abc` and now want to restore the original memory format. The thing is that the original `acb` format has nothing to do with the reordered one (at least conceptually).  So I see no reason for the memory format to _remember_ from what layout it was originally reordered.
In other words the relationship between `aCBde16b16c <-> acbde` is quite the same as between `aCBde16b16c <-> abcde`.

So, I am still puzzled... Could you please elaborate on what is use-case for the function you described?

**Comment by vpirogov** [comment]
I would love to see the mystery resolved eventually. Closing for now.

**Comment by pinzhenx** [comment]
@emfomenk 

I found that `dnnl_md2fmt_str` sorts the strides to infer the permutation order. But for the `dnnl_aCBde16b16c` case above, does that mean this function cannot distinguish the order of the first three dimensions?

**Comment by emfomenk** [comment]
Hi @pinzhenx,

> But for the dnnl_aCBde16b16c case above, does that mean this function cannot distinguish the order of the first three dimensions since they have the same stride 8192?

Correct. `dnnl_md2fmt_str()` is purely a debug function. So even though it can print `aBC...` or `BaC...` for this particular case it doesn't matter a lot: for these sizes these would be equivalent physical formats.

And that's exactly what confuses me -- I cannot see how `aBC...` is different from `BaC...` in this case, i.e. what kind of code might be broken if it guesses the format *incorrectly*. Maybe if you have some kind of map that maps format tag to something, that might bring some problems. But if the code looks at the physical memory format only (i.e. the stride, etc... rather than at the tag) there should be no problem.

**Comment by pinzhenx** [comment]
Hi @emfomenk 
Here's example not directly related to this question.

In the concat it's recommended to have a uniform format for all inputs. Let's say I'm going to align all inputs to one input which has a non-plain format. So how can I implement such a function `dnnl_memory_desc_init_like(dims, type, other_md)` to init a mem desc which has the same layout as another one while having different dimensions?

**Comment by pinzhenx** [comment]
Thanks @emfomenk.
I really appreciate your taking the time to these questions. You've just solved all the myteries about memory desc. Now this issue can actually be closed.
